software/ros_ws/src/perseus_sensors/src/m2m2_lidar/m2m2_lidar.cpp

```diff
@@ -670,6 +554,183 @@ void M2M2Lidar::_readSensorData()
     _scanPublisher->publish(scan);
 }
 
+void M2M2Lidar::_readImuData()
+{
+    RCLCPP_DEBUG(this->get_logger(), "Requesting IMU data...");
+
+    if (!_sendJsonRequest(IMU_COMMAND))
+    {
+        RCLCPP_ERROR(this->get_logger(), "Failed to send IMU data request");
+        return;
+    }
+
+    auto response = _receiveJsonResponse();
+    if (response.empty() || !response.contains("result"))
+    {
+        RCLCPP_ERROR(this->get_logger(), "Failed to receive valid IMU response");
+        return;
+    }
+
+    // Create and populate IMU message
+    auto imu_msg = std::make_unique<sensor_msgs::msg::Imu>();
+    imu_msg->header.stamp = this->now();
+    imu_msg->header.frame_id = this->get_parameter("imu_frame_id").as_string();
+
+    // Extract data from response
+    const auto& result = response["result"];
+
+    // Set orientation quaternion
+    if (result.contains("quaternion"))
+    {
+        imu_msg->orientation.w = result["quaternion"]["w"];
+        imu_msg->orientation.x = result["quaternion"]["x"];
+        imu_msg->orientation.y = result["quaternion"]["y"];
+        imu_msg->orientation.z = result["quaternion"]["z"];
+    }
+
+    // Set angular velocity from gyro
+    if (result.contains("gyro"))
+    {
+        imu_msg->angular_velocity.x = result["gyro"]["x"];
+        imu_msg->angular_velocity.y = result["gyro"]["y"];
+        imu_msg->angular_velocity.z = result["gyro"]["z"];
+    }
+
+    // Set linear acceleration
+    if (result.contains("acc"))
+    {
+        imu_msg->linear_acceleration.x = result["acc"]["x"];
+        imu_msg->linear_acceleration.y = result["acc"]["y"];
+        imu_msg->linear_acceleration.z = result["acc"]["z"];
+    }
+
+    // Set covariance matrices to unknown
+    std::fill(imu_msg->orientation_covariance.begin(),
+              imu_msg->orientation_covariance.end(), -1);
+    std::fill(imu_msg->angular_velocity_covariance.begin(),
+              imu_msg->angular_velocity_covariance.end(), -1);
+    std::fill(imu_msg->linear_acceleration_covariance.begin(),
+              imu_msg->linear_acceleration_covariance.end(), -1);
+
+    _imuPublisher->publish(std::move(imu_msg));
+}
+
+std::vector<std::tuple<float, float, bool>> M2M2Lidar::_interpolatePoints(
+    const std::vector<std::tuple<float, float, bool>>& input_points)
+{
+    using namespace std;
+
+    if (input_points.empty())
+        return {};
+
+    // First normalize and sort points by angle
+    vector<tuple<float, float, bool>> points;
+    points.reserve(input_points.size());
+
+    // Normalize angles to [0, 2π]
+    for (const auto& point : input_points)
+    {
+        float angle = get<0>(point);
+        // Normalize angle to [0, 2π]
+        while (angle < 0) angle += 2 * M_PI;
+        while (angle >= 2 * M_PI) angle -= 2 * M_PI;
```

Comment:
```suggestion
        const auto tau = 2 * std::numbers::pi;
        angle -= std::trunc(angle / tau) * tau;
        if (angle < 0) angle += tau;
```
While loops to process any kind of number are a bad idea, and especially so for floats.

If you use `std::numbers::pi` (yay C++20!) instead of `M_PI` from `<cmath>` you'll need to `#include <numbers>` too